[WIP] Multiprocessing Feature

abdelhakeem opened the pull request:
# GSoC 2019 - Multiprocessing Feature

## Overview

The job control feature of Neovim introduced great multitasking support that added new capabilities and enhanced important tasks by allowing them to be done asynchronously (e.g.: online linting, as in [ALE](https://github.com/dense-analysis/ale) for instance). This improved the overall user experience and reduced editing latency.

We aim to add multiprocessing support to VimL, the built-in scripting language of Neovim, by allowing functions to be invoked asynchronously in separate processes and enabling the retrieval of their results through callbacks and by waiting for them on demand in the parent process.
This multiprocessing feature will have a noticeable positive impact on the performance of parallelizable computations and tasks.

This is achieved by introducing the following built-in multiprocessing primitives based on the job control feature:
- `call_async({callee}, {args}[, {opts}])`
- `call_parallel({callee}, {arglists}[, {opts}])`
- `call_wait({ids}[, {timeout}])`
- `:&:{cmd}` (invokes `{cmd}` asynchronously; has a custom predefined handler for `:[l]vimgrep[add]` commands)

Complementing the multiprocessing feature, is the context feature, which is a set of built-in functions to save and restore parts of the current editor state.

## Progress

- [x] `call_async({callee}, {args}[, {opts}])`
- [x] `call_parallel({callee}, {arglists}[, {opts}])`
- [x] Invoke functions from within a closure in the child process. This ensures that any `l:` variables packed with the sent context are loaded correctly and retained for later requests in case of `call_parallel`.
- [x] Pack scope variables in the sent context when the invoked function is a closure.
- [x] `call_wait({ids}[, {timeout}])`
- [x] `:&:{cmd}` (see https://github.com/neovim/neovim/pull/9943#issuecomment-494728852 and https://github.com/neovim/neovim/pull/9943#issuecomment-518372997)
- [x] `:&:[l]vimgrep[add]...` (see https://github.com/neovim/neovim/pull/9943#issuecomment-497986401 and https://github.com/neovim/neovim/pull/9943#issuecomment-499856486)
- [x] Context feature (see https://github.com/neovim/neovim/pull/10385, https://github.com/neovim/neovim/pull/10430, https://github.com/neovim/neovim/pull/10619, https://github.com/neovim/neovim/pull/10637)
  - [x] Context stack and `ctx`-family functions
  - [x] `nvim_get_context({types})` and `nvim_load_context({ctx})`
  - [x] Save and restore registers
  - [x] Save and restore jump list
  - [x] Save and restore buffer list
  - [x] Save and restore global variables
  - [x] Save and restore script-local variables
  - [x] Save and restore buffer-local, window-local, and tabpage-local variables
  - [x] Save and restore function-local variables
  - [x] Save and restore parent-scope variables for closures
  - [x] Save and restore global functions
  - [x] Save and restore script-local functions
  - [ ] More...?
- [x] Documentation
- [x] More test coverage

abdelhakeem committed:
multiproc: add preliminary async call support

abdelhakeem committed:
api: add nvim_grep method

justinmk commented:
nice progress! 

I think it makes sense to implement parts in Lua where possible. That will avoid stuff like [this](https://github.com/neovim/neovim/pull/9943/files#diff-deb695d507653839019814f54242d47bR7141).

See [nvim_get_proc_children](https://github.com/neovim/neovim/blob/0be8fb47a46ee472af41dd96c5884e8e69b94639/src/nvim/api/vim.c#L2066-L2067) for example. Most of its logic is implemented [here](https://github.com/neovim/neovim/blob/0be8fb47a46ee472af41dd96c5884e8e69b94639/src/nvim/lua/vim.lua#L38). 

Similar approach can be done for this work. Just start adding functions in `lua/vim.lua` as needed.

abdelhakeem commented:
UPDATE: now supports callbacks and keeps async call jobs (up to 10, number is subject to discussion) in a pool to reuse them on later calls.

Here is a simple demo that demonstrates the feature in its current state:
```vimscript
function! HandleNvimGrepResults(results) abort
  for result in a:results
    call nvim_command('badd '.result.fname)
    let result.bufnr = bufnr(result.fname)
  endfor
  call setqflist(a:results)
  copen
endfunction
```

![](https://media.giphy.com/media/kGFMYzpYSFYYlO13Ex/giphy.gif)

I'll refactor some code before moving on, probably reimplementing `new_async_call_job` and `nvim_grep` mostly in Lua. Any similar stuff to follow will be implemented in Lua for quicker and cleaner progress.

justinmk commented on src/nvim/eval.c:
```diff
@@ -7076,6 +7076,70 @@ static void f_assert_true(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 }
 
+/// "async_invoke(callee, args, opts[, callback])" function
+static void f_async_invoke(typval_T *argvars, typval_T *rettv, FunPtr fptr)
```
Since there is already a `call()` function (yes, `:call call(...)`) I guess it makes sense to name this `call_async()`.

justinmk commented on src/nvim/channel.h:
```diff
@@ -130,5 +140,28 @@ static inline Stream *channel_outstream(Channel *chan)
   abort();
 }
 
+static inline Channel *acquire_asynccall_channel(void)
```
as discussed in chat, I think we should avoid managing a pool. On a warm system spawning a new `nvim` costs ~10 milliseconds (as long as it doesn't load user config).

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -187,6 +187,34 @@ deepcopy = function(orig)
+local function _create_nvim_job()
+  local progpath = vim.api.nvim_get_vvar('progpath')
+  return vim.api.nvim_call_function('jobstart', {
+    { progpath, '--embed', '--headless', '-u', 'NONE', '-i', 'NONE' },
```
`-n` is probably a good idea also.

abdelhakeem committed:
multiproc: add call_wait function

abdelhakeem commented:
UPDATE: implemented `call_wait`
Example usage:
```vim
" Searches "paths" for "pattern" in parallel
function! MultiGrep(pattern, paths, global)
  let jobs = []
  for path in a:paths
    let job = call_async('nvim_grep', [a:pattern, path, a:global], {})
    call add(jobs, job)
  endfor
  let results_list = call_wait(jobs)
  let results = []
  for result in results_list
    call extend(results, result.value)
  endfor
  return results
endfunction
```

justinmk commented:
Next step is to write a basic test, see `test/functional/example_spec.lua` . The usage example you gave could be that basic test, it could search some fake directory created by the test.

Then maybe it's time to think about how `:vimgrep` could be invoked like this more conveniently. I'm thinking we could special-case the `:&` command.

- `:&:foo` (notice the `:` preceding `foo`) invokes the "async handler" of `:foo` .
- If a command doesn't have an "async handler", it's  an error (for now). 
- `:&foo` (without `:` preceding `foo`) acts like the old `:&` command, `foo` is just passed as flags to `:&`.

Thus a user could async-invoke `:vimgrep` like this:

    :&:vimgrep /pattern/ *

1. The async handler for `:vimgrep` divides the `*` pattern into some number of paths
2. Does  `call_wait(nvim_grep,...)`, collects the results
3. After results are collected, they are put into the quickfix list

justinmk commented on src/nvim/eval.c:
```diff
@@ -7419,6 +7419,105 @@ static void f_call(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+}
+
+/// "call_wait(ids[, timeout])" function
+static void f_call_wait(typval_T *argvars, typval_T *rettv, FunPtr fptr)
```
This seems like another case where we could avoid a bunch of C code in favor of Lua. Perhaps the `job_results` list should be a Lua table, then it's easier to work with.

abdelhakeem committed:
multiproc: add multiproc_spec.lua

abdelhakeem committed:
api: add nvim_grep test

abdelhakeem commented:
UPDATE: implemented a new command (`&:`) as proposed in @justinmk's comment and implemented its `vimgrep` handler, the results look promising...

```vim
" experiments/async_vimgrep.vim
let start = reltime()
vimgrep /buf_T/jg **/*
echom 'vimgrep took '.reltimestr(reltime(start)).' seconds'

let start = reltime()
&:vimgrep /buf_T/g **/*
echom '&:vimgrep took '.reltimestr(reltime(start)).' seconds'
```

```vim
source experiments/async_vimgrep.vim
```

![Comparison](https://user-images.githubusercontent.com/20467669/58754828-c0905e80-84d8-11e9-8248-7d96e99c49d4.png)

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -2423,9 +2423,14 @@ static char_u *find_command(exarg_T *eap, int *full)
+      } else if (vim_strchr((char_u *)"@!=><&~#", *p) != NULL) {
+        p++;
+      }
+    }
```
Other way to handle ambiguity:

- add a new function like `ex_sub_or_async` which decides whether to call `ex_substitute` or `ex_async`. 
  -  use `skipwhite()`  to check if the next char is `:`. 
- in `ex_cmds.lua`, 
  - set `func='ex_sub_or_async'` for the `&` command.
  - add `NOTRLCOM`


Not sure if that's better, can just leave it for now.

abdelhakeem replied:
Hmmm...I didn't want to change the `&` command mainly because semantically our new async command is not a special case of substitution.

Concerning `&:vimgrep` handler, I need to modify it so that it can handle paths more properly just like `vimgrep`, currently it just search `glob(path, v:true, v:true)` of whatever path given, so paths like `**/*.h **/*.c` will not work unlike `vimgrep`, not a big deal to parse I guess.

Not sure about the next step, maybe I'll add similar async equivalents to `vimgrepadd`, `lvimgrep`, and `lvimgrepadd` for now.

I'm open to any kind of feedback, once we settle we can start working on the more interesting stuff like context handling and async calls of user-defined functions (which could be just making the called function part of the sent context and then call it as usual).

justinmk replied:
> Not sure about the next step, maybe I'll add similar async equivalents to `vimgrepadd`, `lvimgrep`, and `lvimgrepadd` for now.

Yes, implementing a few more cases will help to start seeing a pattern (and see what's painful/boilerplate/brittle). Then we can start to form a best-practice for adding this functionality.

abdelhakeem commented:
UPDATE: implemented "async handlers" for the rest of `vimgrep` commands (`vimgrepadd`, `lvimgrep`, etc...) and modified the argument parsing logic for patterns and paths, hopefully it matches that of the original commands now.

```vim
function! s:profile_cmd(cmd) abort
  echom 'executing: '.a:cmd
  let start = reltime()
  silent! call execute(a:cmd)
  let elapsed = reltimefloat(reltime(start))
  echom len(getqflist()).' matches in '.string(elapsed).' seconds'
  return elapsed
endfunction

function! CompareCmds(cmd1, cmd2) abort
  let elapsed1 = s:profile_cmd(a:cmd1)
  let elapsed2 = s:profile_cmd(a:cmd2)
  echom 'speedup: '.string(elapsed1 / elapsed2)
endfunction
```

```vim
call CompareCmds('vimgrep /buf_T/jg **/*.c **/*.h', '&:vimgrep /buf_T/jg **/*.c **/*.h')
```

![image](https://user-images.githubusercontent.com/20467669/59102142-7b7e8900-892b-11e9-913d-01dc23f1af98.png)

asymmetric commented:
I think you meant `elapsed1 - elapsed2` :)

abdelhakeem commented:
> I think you meant `elapsed1 - elapsed2` :)
- @asymmetric 

Not really, it's like _how many times faster_ `cmd2` is. For this particular example the result is usually around 20.

abdelhakeem committed:
multiproc: add &: command

abdelhakeem committed:
multiproc: add vimgrep-family async handlers

abdelhakeem committed:
api: add nvim_grep doc

abdelhakeem committed:
multiproc: add v:cores

abdelhakeem committed:
multiproc: add call_parallel

abdelhakeem commented:
UPDATE: added `call_parallel`, which implements the common workers pool pattern where a specified number of worker processes is spawned and fed work items, keeping them all busy, until the work queue is exhausted. It's somehow similar to Python's [Pool](https://docs.python.org/3/library/multiprocessing.html#using-a-pool-of-workers) class.

```vim
" call_parallel(count, callee, arglists, opts[, callback])

function! MultiGrep(paths, pattern, global, ...) abort
  let args = map(a:paths, '[a:pattern, v:val, a:global]')
  if (a:0 > 0)
    return call_parallel(v:cores, 'nvim_grep', args, {}, a:1)
  else
    return call_parallel(v:cores, 'nvim_grep', args, {})
  endif
endfunction
```

This will have more interesting uses when we introduce multiprocessing support to user-defined functions as well.

blueyed commented on runtime/doc/eval.txt:
```diff
@@ -2826,6 +2839,74 @@ call({func}, {arglist} [, {dict}])			*call()* *E699*
+			btext	  include buffer text
+
+							*call_parallel()*
+call_parallel({count}, {func}, {argslist}, {opts}[, {callback}])
```
Re API `count` could go into `opts` and default to all cores (`v:cores`) maybe?

abdelhakeem replied:
Hmmm...yeah this is a sane default

abdelhakeem commented:
Now that the Context feature is merged (https://github.com/neovim/neovim/pull/10637 and https://github.com/neovim/neovim/pull/10619), although more work will follow later on, we can continue working on this PR. I'll push some changes soon including few interface changes (e.g.: https://github.com/neovim/neovim/pull/9943#discussion_r300838163) and integration with the context feature.

abdelhakeem committed:
multiproc: integrate context feature

abdelhakeem commented:
These two unrelated tests almost always fail:
```
[ RUN      ] 'wildmenu' does not crash after cycling back to original text: ERR
test/functional/ui/screen.lua:579: Row 4 did not match.
Expected:
  |                         |
  |~                        |
  |~                        |
  |*!  #  &  <  =  >  @  >   |
  |:!^                       |
Actual:
  |                         |
  |~                        |
  |~                        |
  |*!  #  &  &:  <  =  >  >  |
  |:!^                       |
To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  
stack traceback:
	test/functional/ui/screen.lua:579: in function '_wait'
	test/functional/ui/screen.lua:367: in function 'expect'
	test/functional/ui/wildmode_spec.lua:74: in function <test/functional/ui/wildmode_spec.lua:62>

[ RUN      ] 'wildmenu' with laststatus=0, :vsplit, :term #2255: ERR
test/functional/ui/screen.lua:579: Failed to match any screen lines.
Expected (anywhere): "!  #  &  <  =  >  @  >   |
:!^"
Actual:
  |$                   │    |
  |                    │~   |
  |                    │~   |
  |!  #  &  &:  <  =  >  >  |
  |:!^                       |
stack traceback:
	test/functional/ui/screen.lua:579: in function '_wait'
	test/functional/ui/screen.lua:367: in function 'expect'
	test/functional/ui/wildmode_spec.lua:21: in function 'expect_stay_unchanged'
	test/functional/ui/wildmode_spec.lua:165: in function <test/functional/ui/wildmode_spec.lua:140>
```

blueyed commented:
> These two unrelated tests almost always fail in Travis:

Good to know / interesting.  They are not flaky otherwise (that much at least), so something here might trigger it.  Not necessarily a problem with the code, and the tests likely need fixing, but just for info.

abdelhakeem committed:
multiproc: fix memory leak

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -2354,3 +2351,113 @@ Array nvim__inspect_cell(Integer grid, Integer row, Integer col, Error *err)
+  ADD(args, STRING_OBJ(path));
+  ADD(args, BOOLEAN_OBJ(global));
+  Object results = nvim_execute_lua(
+      STATIC_CSTR_AS_STRING("return vim._grep(select(1, ...))"), args, err);
```
`select(1, ...)` is redundant, can be just `return vim._grep(...)`.

bfredl commented on src/nvim/channel.h:
```diff
@@ -137,5 +153,41 @@ static inline Stream *channel_outstream(Channel *chan)
+  ADD(args, INTEGER_OBJ((long)job));
+  ADD(args, copy_object(result));
+  nvim_execute_lua(
+      STATIC_CSTR_AS_STRING("return vim._put_result(select(1, ...))"),
```
Whenever the return value of (c function) `nvim_execute_lua` is not used, we should drop `return ` from the lua statement. This ensures an allocated API value cannot be leaked, even if the lua code did something unexpected.

bfredl commented on src/nvim/channel.h:
```diff
@@ -137,5 +153,41 @@ static inline Stream *channel_outstream(Channel *chan)
+  ADD(args, INTEGER_OBJ((long)job));
+  ADD(args, copy_object(result));
+  nvim_execute_lua(
+      STATIC_CSTR_AS_STRING("return vim._put_result(select(1, ...))"),
```
Whenever the return value of (c function) `nvim_execute_lua` is not used, we should drop `return ` from the lua statement. This ensures an allocated API value cannot be leaked, even if the lua code did something unexpected.

bfredl replied:
also, if `nvim_execute_lua(STATIC_CSTR_AS_STRING("..."), ...)` becomes common (which is a good thing), maybe it can be a dedicated macro: `EXEC_LUA_STATIC("vim._put_result(...)", args, err)`

abdelhakeem committed:
multiproc/API: add EXEC_LUA_STATIC macro

- Added EXEC_LUA_STATIC in nvim/api/vim.h
- Updated vimgrep-family commands async handler to use the new
  `call_parallel` interface

abdelhakeem committed:
multiproc: update docs

abdelhakeem committed:
fixup! multiproc: integrate context feature

call_parallel: no need to reload context on subsequent RPC calls to the
same job.

abdelhakeem committed:
fixup! multiproc/API: add EXEC_LUA_STATIC macro

abdelhakeem committed:
fixup! multiproc: add multiproc_spec.lua

abdelhakeem committed:
multiproc: add support for user-defined functions

abdelhakeem committed:
multiproc: update multiproc_spec.lua

abdelhakeem commented:
[UPDATE]

Now supports user-defined functions, including script-local functions, funcrefs, and lambda expressions.

Interesting use case in the specs:
```lua
nvim('set_var', 'A',  { { -1,  0,  0,  0 },
                        {  0, -1,  0,  0 },
                        {  0,  0, -1,  0 },
                        {  0,  0,  0, -1 } })

nvim('set_var', 'B', { { 1, 2, 3, 4 },
                       { 5, 6, 7, 8 },
                       { 1, 2, 3, 4 },
                       { 5, 6, 7, 8 } })

source([[
function CalculateElement(i, j)
  let value = 0
  let b_idx = 0
  for a in g:A[a:i]
    let value += a * g:B[b_idx][a:j]
    let b_idx += 1
  endfor
  return [a:i, a:j, value]
endfunction
]])

-- Prepare arguments and result (all zeros)
local result = {}
local indices = {}
for i = 0,3 do
  table.insert(result, {})
  for j = 0,3 do
    table.insert(result[i+1], 0)
    table.insert(indices, {i, j})
  end
end
nvim('set_var', 'Result', result)

source([[
function Retrieve(r)
  for e in a:r
    let g:Result[ e[0] ][ e[1] ] = e[2]
  endfor
endfunction
]])

local jobs = call('call_parallel', 'CalculateElement', indices,
                  { count = 2,
                    done = 'Retrieve',
                    context = nvim('get_context', {'gvars'})})
call('call_wait', jobs)
eq({ { -1, -2, -3, -4 },
     { -5, -6, -7, -8 },
     { -1, -2, -3, -4 },
     { -5, -6, -7, -8 } }, nvim('get_var', 'Result'))
```

One caveat is about using script-local functions inside the body of the called function. For example, this code will trigger the following error (despite the presence of `sfuncs` in the context):

```vim
function s:greet(name)
  return 'Hello, '.a:name.'!'
endfunction

function Greet(name)
  return s:greet(a:name)
endfunction

call call_async('Greet', ['Neovim'], { 'context': nvim_get_context(['sfuncs']) })

```

```
multiproc: job 3: Vim(return):E120: Using <SID> not in a script context: s:greet
```

abdelhakeem commented:
Some unnreleated unit tests are failing [here](http://neovim-qb.szakmeister.net/build/23935/tap_report/by_test?reportset=unittest-freebsd-64-debug), no idea what triggers them...

justinmk commented:
Ignore quickbuild for now, particularly those multibyte unittests.

abdelhakeem committed:
fixup! api: add nvim_grep method

abdelhakeem committed:
fixup! multiproc: add call_wait function

abdelhakeem committed:
fixup! multiproc: add support for user-defined functions

abdelhakeem committed:
fixup! multiproc: add support for user-defined functions

abdelhakeem committed:
fixup! multiproc: update multiproc_spec.lua

abdelhakeem committed:
fixup! multiproc: update docs

abdelhakeem committed:
context: return decoded values in context dict

abdelhakeem committed:
context: better error handling

abdelhakeem committed:
multiproc: more coverage

abdelhakeem committed:
multiproc: call funcs in script context (if any)

Invoke async functions in the script context in which they are
defined (if any). This allows async-called functions to use
script-local variables and functions, provided that the
corresponding context item is included in the sent context.

abdelhakeem committed:
fixup! multiproc: more coverage

abdelhakeem committed:
multiproc: fix memory leaks

abdelhakeem committed:
fixup! multiproc: more coverage

abdelhakeem committed:
multiproc: introduce &: default async handler

- introduce &: default async handler
- refurbish &: vimgrep async handler

abdelhakeem committed:
fixup! multiproc: introduce &: default async handler

abdelhakeem committed:
multiproc: add async_handler_spec.lua

abdelhakeem commented:
[UPDATE]

1. Asynchronously invoked functions are now called in a script context (if any) with the same `SID` value of the parent. This allows script-local variables (not yet added to context) and functions to be referenced in the called function, given that the corresponding context items are included in the sent context (https://github.com/neovim/neovim/pull/9943/commits/f5e4a3632c5bda4a2978a0c48b53cbde7e9405de). Example:

```vim
function s:greet(name)
  return 'Hello, '.a:name.'!'
endfunction

echo call_wait([call_async({ name -> s:greet(name) }, ['Neovim'],
                         \ { 'context': nvim_get_context(['sfuncs']) })])
```

2. Default `&:` command async handler is introduced. Overview:
```c
/// ":&:{cmd}" command
///
/// Invokes the async handler of "{cmd}" (i.e.: a function that invokes
/// "{cmd}" in separate process(es) and handles the results
/// in a pre-defined way specific to each command.)
///
/// If "{cmd}" does not have an async handler, a default handler that does
/// the following is invoked:
///
///   1. Asynchronously run the command in a separate process with a
///      snapshot of the current context (via nvim_get_context([v:null])).
///   2. Re-emit command output in the parent (i.e.: calling) process.
///   3. Load the (possibly modified) context returned by the child.
```

This automatically provides the following shortcut for `call_async` calls:
```vim
&:call Foo(args...)
```

There are still, however, many edge cases and tricks regarding the default handler. For instance, the following command has no effect on the buffer list of the parent process:
```vim
&:bwipeout 2
```
Buffer 2 will be wiped out from the child process and hence from the received context, except that the context feature in its current state simply "merges" the loaded context, so any removals from the buffer list will not be reflected in the parent.

You can imagine how many such edge cases may exist (e.g.: `&:call...` called functions never run in a script context, due to the way the handler works). So, in my opinion, it'll be important to clearly document how both features (context and default async handler) work.

As a next step, I'm thinking of introducing a special handler for the `call` command to alleviate the previously mentioned issues.

As for the context feature, I'm not quite satisfied about how it "hacks around" the ShaDa module to provide the save/restore functionality, which has side-effects like lack of flexibility with internal representation and context dictionary format. Ideally, the ShaDa module can be modularized and decoupled enough to avoid duplicate or unnecessary work.

justinmk commented:
> There are still, however, many edge cases and tricks regarding the default handler. For instance, `&:bwipeout 2` has no effect on the buffer list of the parent process

That's probably a good thing. Generally side-effects should not affect the parent. The "default handler" should only do non-surprising, non-destructive things: collect/display output, provide some useful data, etc.

> As for the context feature, I'm not quite satisfied ... lack of flexibility with internal representation and context dictionary format.

Let's approach this with concrete examples before thinking about a big refactor. For example, I can imagine that the shada key names are rather unfriendly: so that could be handled by just remapping some of the key names in e.g. `ctx_to_dict()`.

abdelhakeem committed:
fixup! multiproc: introduce &: default async handler

abdelhakeem commented:
> The "default handler" should only do non-surprising, non-destructive things: collect/display output, provide some useful data, etc.

Agreed. Some exceptions might better be modifiable though, like shared context variables (currently only global) and registers, as is the case now. That's currently the only way to get back useful data from this command.

I'm considering implementing a custom handler for `call` to allow callbacks to be specified among other things.

Also, not related to "async handlers", I'm thinking of adding another optional field in the `opts` dictionary of `call_parallel` to specify a per-item callback that gets called with the result of every finished work item, this could be useful and more convenient in some cases, including the matrix multiplication example above.

> Let's approach this with concrete examples before thinking about a big refactor. For example, I can imagine that the shada key names are rather unfriendly: so that could be handled by just remapping some of the key names in e.g. `ctx_to_dict()`.

Additionally, the three fields preceding each entry (type, timestamp, length) are only necessary for ShaDa parsing and reading. I'll consider removing them in `ctx_to_dict()` as well. The type and length field can be recalculated in `ctx_from_dict()` and the timestamp value is not that critical to retain.

abdelhakeem committed:
multiproc: call_parallel: add "itemdone" callback

abdelhakeem committed:
multiproc: move asynccall_* functions to channel.c

The includes required for the `asynccall_*` family of
functions mess up with some unit tests on CI when included
in "nvim/channel.h", so move them to "nvim/channel.c".

abdelhakeem committed:
Context: polish context dictionary format

Make context dictionary more user-friendly:
- Eliminate buffer list unnecessary nesting
- Eliminate (type, timestamp, length) fields

abdelhakeem committed:
fixup! Context: polish context dictionary format

abdelhakeem committed:
fixup! Context: polish context dictionary format

Map register entry key names to more user-friendly names

abdelhakeem committed:
fixup! Context: polish context dictionary format

Validate register entry type (must be dictionary).

abdelhakeem committed:
fixup! Context: polish context dictionary format

More key name mappings.

abdelhakeem committed:
Context: fix bug with decoding large context

- Reload `msgpack_unpacker` buffer in `sbuf_to_array` instead
of failing.
- More helpful error messages in `sbuf_to_array`.

abdelhakeem commented:
[UPDATE]

1. Per-item callback added for `call_parallel`. This is specified by the `"itemdone"` key in the `opts` dictionary. Both types of callbacks (per-item and overall) can also be specified at the same time, enabling the user to do some action upon retrieval of each work item result and another action upon completion of all work items. Revisiting the matrix multiplication use case:

```vim
function! s:zero_square_matrix(dim)
  return map(range(1, a:dim), 'map(range(1, a:dim), 0)')
endfunction

function! s:random_square_matrix(dim)
  return map(range(1, a:dim),
   \         { -> map(range(1, a:dim), 'luaeval("math.random()")') })
endfunction

function! s:calculate_element(i, j)
  let val = 0
  let idx = 0
  for a in g:A[a:i]
    let val += a * g:B[idx][a:j]
    let idx += 1
  endfor
  return [a:i, a:j, val]
endfunction

function! s:position_pairs(dim)
  let pairs = []
  for i in range(a:dim)
    for j in range(a:dim)
      call add(pairs, [i, j])
    endfor
  endfor
  return pairs
endfunction

let g:dim = 100

let g:A = s:random_square_matrix(g:dim)
let g:B = s:random_square_matrix(g:dim)
let s:C1 = s:zero_square_matrix(g:dim)
let s:C2 = s:zero_square_matrix(g:dim)

function! MatrixMultiply()
  for i in range(g:dim)
    for j in range(g:dim)
      for k in range(g:dim)
        let s:C1[i][k] += g:A[i][j] * g:B[j][k]
      endfor
    endfor
  endfor
endfunction

function! ParallelMatrixMultiply()
  call call_wait(call_parallel('s:calculate_element', s:position_pairs(g:dim), {
   \                           'context': nvim_get_context(['gvars']),
   \                           'itemdone': { rv ->
   \                             execute('let s:C2[rv[0]][rv[1]] = rv[2]')
   \                           } }))
endfunction

profile start prof
profile func *MatrixMultiply
call MatrixMultiply()
call ParallelMatrixMultiply()
```

```
FUNCTION  ParallelMatrixMultiply()
Called 1 time
Total time:  10.577885
 Self time:   9.480026
.
.
.
FUNCTION  MatrixMultiply()
Called 1 time
Total time:  27.649642
 Self time:  27.649642
.
.
.
```

2. Context dictionary format redefined in a more user-friendly way:
  - No unnecessary fields (type, timestamp, and length)
  - Clearer key names

Now it's a lot easier to even hardcode some context dictionary by hand or to modify one obtained with `ctx_get()` or `nvim_get_context()`. Example:

```vim
call nvim_load_context({'gvars': [ ['foo', 1], ['bar', [1, 2, 3]] ]})
echo [foo, bar]
"> [1, [1, 2, 3]]

call nvim_load_context({'regs': [ {'name': '0', 'content': ['Zero']} ]})
echo getreg(0)
"> Zero
```

abdelhakeem committed:
multiproc: fix call_parallel memory bugs

abdelhakeem committed:
Context: fix malformed context memory leak

abdelhakeem committed:
Context: check entries in ctx_restore_funcs

abdelhakeem committed:
fixup! Context: polish context dictionary format

abdelhakeem committed:
fixup! Context: polish context dictionary format

Improve key mapping logic

abdelhakeem committed:
fixup! multiproc: fix call_parallel memory bugs

abdelhakeem committed:
fixup! multiproc: fix call_parallel memory bugs

abdelhakeem committed:
multiproc: improve &:vimgrep async handler

- make it non-blocking
- utilize `call_parallel` per-item callback (see c34373eba8)

abdelhakeem committed:
fixup! multiproc: fix call_parallel memory bugs

abdelhakeem committed:
fixup! Context: polish context dictionary format

Do not add empty context items in ctx_to_dict

abdelhakeem committed:
fixup! multiproc: improve &:vimgrep async handler

abdelhakeem committed:
fixup! Context: polish context dictionary format

abdelhakeem committed:
multiproc: count AsyncCall references

abdelhakeem committed:
fixup! multiproc: fix call_parallel memory bugs

abdelhakeem committed:
Context: report more info for errors

abdelhakeem committed:
multiproc: collect &:vimgrep results to save mem

abdelhakeem committed:
fixup! multiproc/API: add EXEC_LUA_STATIC macro

abdelhakeem committed:
Context: improve context function handling

- Invoke functions with current_SID value
- Include uf_script_ID function values
- Improve error messages

abdelhakeem committed:
fixup! Context: improve context function handling

abdelhakeem committed:
Context/multiproc: save/restore script-local vars

abdelhakeem commented:
[UPDATE]

Besides some memory bug fixes, script-local functions are now saved/restored with the `current_SID` value in which they are defined so that they remain script-local and can refer to script-local variables as expected.

The interesting update is the introduction of script-local variables to the context feature, which eases some use cases for plugins willing to utilize the multiprocessing feature (i.e.: sharing "private state" with child workers). For instance:

```vim
function! s:zero_square_matrix(dim)
  return map(range(1, a:dim), 'map(range(1, a:dim), 0)')
endfunction

function s:position_pairs(dim)
  let pairs = []
  for i in range(a:dim)
    for j in range(a:dim)
      call add(pairs, [i, j])
    endfor
  endfor
  return pairs
endfunction

function s:calculate_element(i, j)
  let val = 0
  let idx = 0
  for a in s:A[a:i]
    let val += a * s:B[idx][a:j]
    let idx += 1
  endfor
  return [a:i, a:j, val]
endfunction

function! RandomSquareMatrix(dim)
  return map(range(1, a:dim),
   \         { -> map(range(1, a:dim), 'luaeval("math.random()")') })
endfunction

function! ParallelMatrixMultiply(A, B)
  let dim = len(a:A)
  let s:result = s:zero_square_matrix(dim)
  let s:A = a:A
  let s:B = a:B
  call call_wait(call_parallel('s:calculate_element', s:position_pairs(dim), {
   \                           'context': nvim_get_context(['svars']),
   \                           'itemdone': { rv ->
   \                             execute('let s:result[rv[0]][rv[1]] = rv[2]')
   \                           } }))
  return s:result
endfunction
```

Now, this can be transparently invoked by the user like:
```vim
let myresult = ParallelMatrixMultiply(mybigmatrix1, mybigmatrix2)
```

abdelhakeem committed:
Context: handle sandboxed functions

abdelhakeem committed:
eval/multiproc: define ISLAMBDA macro

abdelhakeem committed:
Context: save/restore b:, w:, and t: vars

abdelhakeem committed:
Context: save/restore l: vars

abdelhakeem committed:
fixup! multiproc: fix call_parallel memory bugs

abdelhakeem committed:
Context: save/restore parent-scope for closures

abdelhakeem committed:
fixup! Context: handle sandboxed functions

abdelhakeem commented:
[UPDATE]

I've recently added more items to the context feature, namely buffer-local, window-local, and tabpage-local variables, in addition to function-local and parent-scope variables for closures.
I've included the next steps in the description of this PR.

cc @justinmk

abdelhakeem committed:
multiproc: invoke function from within a closure

Invoke functions from within a closure in the child process. This ensures that any `l:` variables packed with the sent context are loaded correctly and retained for later requests in case of `call_parallel`.

abdelhakeem committed:
fixup! multiproc: invoke function from within a closure

justinmk commented:
> I've recently added more items to the context feature, namely buffer-local, window-local, and tabpage-local variables, in addition to function-local and parent-scope variables for closures.

Really nice! We will eventually want "options" in Context too, but that can wait.

abdelhakeem committed:
multiproc: pack parent-scope l: vars for closures

abdelhakeem committed:
multiproc: set v:jobs in callbacks

- in "done" callback, v:jobs is a list of all call jobs (one element in case of `call_async`)
- in "itemdone" callback, v:jobs is a list with one element: the id of the job that completed the corresponding work item

abdelhakeem committed:
multiproc: reap async handler jobs

abdelhakeem committed:
multiproc: fix call_parallel memory leak

abdelhakeem committed:
multiproc: improve &:vimgrep handler

abdelhakeem committed:
multiproc: update docs

abdelhakeem commented:
This concludes our GSoC project, but not my work on this feature!
It has been a great experience and much has been done so far.

Next steps for now:
- More coverage
- More examples in documentation
- Context additions
- Context documentation

All work to follow will be in #10843. Stay tuned!
